File: platform/src/main/java/fun/fengwk/kkstudio/platform/catalog/mcp/service/McpServerService.java
package fun.fengwk.kkstudio.platform.catalog.mcp.service;

import fun.fengwk.convention4j.api.page.Page;
import fun.fengwk.convention4j.api.page.PageQuery;

import fun.fengwk.kkstudio.share.ai.mcp.McpServerConfigDTO;
import fun.fengwk.kkstudio.share.ai.mcp.McpServerCreateDTO;
import fun.fengwk.kkstudio.share.ai.mcp.McpServerDTO;
import fun.fengwk.kkstudio.share.ai.mcp.McpServerUpdateDTO;

/** Platform MCP server 应用服务：name-keyed CRUD/CAS、显式发现与引用保护流程。 */
public interface McpServerService {

  Page<McpServerDTO> pageServers(PageQuery pageQuery);

  /** 按不可变 name 获取 Server 安全公开视图。 */
  McpServerDTO getServer(String name);

  /** 读取 Server 显式 HTTP 配置；headers 可能内嵌凭据，调用方必须禁止缓存。 */
  McpServerConfigDTO getServerConfig(String name);

  /** 创建 Server 配置：仅校验并持久化为 UNVERIFIED，不发起任何网络 I/O。 */
  McpServerDTO createServer(McpServerCreateDTO createDTO);

  /** 基于 (name, expectedVersion) 的 CAS 全量更新：仅更新配置并重置状态为 UNVERIFIED，不发起发现。 */
  McpServerDTO updateServer(String name, McpServerUpdateDTO updateDTO);

  /** 基于 (name, expectedVersion) 的显式发现：同步握手并原子替换当前工具行。 */
  McpServerDTO discoverServer(String name, String expectedVersion);

  /** 基于 (name, expectedVersion) 的硬删除 CAS：被 agent 引用时拒绝删除。 */
  void deleteServer(String name, String expectedVersion);
}

File: frontend/src/shared/api/mcp-server-service.ts
import { apiClient, type HttpClient } from '@/shared/api/client'
import type {
  McpServerConfigDTO,
  McpServerCreateDTO,
  McpServerDTO,
  McpServerUpdateDTO,
} from '@/shared/api/contracts/ai-mcp'
import type { PageResult } from '@/shared/api/contracts/base'

export function createMcpServerService(client: HttpClient = apiClient) {
  return {
    pageServers: (pageNumber = 1, pageSize = 50): Promise<PageResult<McpServerDTO>> =>
      client.get('/ai/mcp-servers', { params: { pageNumber, pageSize } }),

    getServer: (name: string): Promise<McpServerDTO> =>
      client.get(`/ai/mcp-servers/${encodeURIComponent(name)}`),

    /**
     * 按需读取当前完整配置（含 URL 与 headers，强制 Cache-Control: no-store）。
     * 仅在显式编辑时直读，绝不进入 React Query、LocalStorage 或 URL。
     */
    getServerConfig: (name: string): Promise<McpServerConfigDTO> =>
      client.get(`/ai/mcp-servers/${encodeURIComponent(name)}/config`, {
        headers: { 'Cache-Control': 'no-store' },
      }),

    createServer: (data: McpServerCreateDTO): Promise<McpServerDTO> =>
      client.post('/ai/mcp-servers', data),

    updateServer: (name: string, data: McpServerUpdateDTO): Promise<McpServerDTO> =>
      client.put(`/ai/mcp-servers/${encodeURIComponent(name)}`, data),

    /**
     * 触发 MCP Server 工具发现（POST /ai/mcp-servers/{name}/discover?expectedVersion=...）。
     */
    discoverServer: (
      name: string,
      expectedVersion: string,
    ): Promise<McpServerDTO> =>
      client.post(
        `/ai/mcp-servers/${encodeURIComponent(name)}/discover?expectedVersion=${encodeURIComponent(expectedVersion)}`,
      ),

    deleteServer: (name: string, expectedVersion: string): Promise<void> =>
      client.delete(
        `/ai/mcp-servers/${encodeURIComponent(name)}?expectedVersion=${encodeURIComponent(expectedVersion)}`,
      ),
  }
}

export const mcpServerService = createMcpServerService()
